### PR TITLE
use templated TPlayFactory class instead of REGISTER_PLAY macro

### LINKED ISSUE #35: Investigate the use of templating for the REGISTER_PLAY macro
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
Macros can be tricky to follow and leave opportunity for bugs that the compiler can't always catch. It might be better to use templating in favor of the REGISTER_PLAY macro (if that's possible).

**Describe the solution you'd like**
The PlayFactory class could be templated, and then we could simply initialize a static variable instead of using the macro. The static variable would accept the class of whatever Play is being implemented (eg. MovePlay) as the type. This would then give us compile-time guarantees and allow us to get rid of the macro.

Example code from @garethellis0 , from #32 
```
template <typename PLAY_T>
class PlayFactory {
    public :
        std::shared_ptr<PlayFactory> getFactoryInstance()
        {
            if (!factory_instance){
                factory_instance = std::make_shared<PlayFactory<PLAY_T>>;
                Play::registerPlay(std::shared_ptr<PlayFactory<PLAY_T>>(
                            *factory_instance)); 
            }
            return factory_instance
        }

        std::shared_ptr<Play> getInstance() 
        {
            return std::make_shared<PLAY_T>();
        }

    private:
        static std::shared_ptr<PlayFactory<PLAY_T>> factory_instance;

        // Default constructor is private to force caller to use 
        // getFactoryInstance
        PlayFactory();

}
```

**Additional context**
I (Mathew) took a quick look at this before. Creating the simple template seems straightforward but it looks like the issues will come from nitty-gritty ordering of class declarations and other similar issues. For example, using a forward delcaration of the PlayFactory causes the implementation below to warn `redefinition of 'PlayFactory' as a different kind of symbol`

### PR DESCRIPTION
**Pull request by jordanzliu**
resolves #35

### CODE REVIEW

File: src/thunderbots/software/ai/hl/stp/play/play.h [hunk 1]
```diff
@@ -136,33 +136,36 @@ class Play
...
- * macro).
+ * templated class).
  *
- * @param play_class_name The classname of the Play to be added to the registry. For
- * example, to add new class called MovePlay that inherits from Play, the following line
- * should be added to the end of the .cpp file (without the quotations):
- * "REGISTER_PLAY(MovePlay)"
+ * @param <T> The class of the Play to be addred to the registry. For example, to add a
+ * new class called MovePlay that inherits from Play, the following line should be added
+ * to the end of the .cpp file (without the quotations):
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/play/play.h]: nit; but looks like templated params should be `@tparam` https://stackoverflow.com/questions/13359217/how-to-document-c-templates-and-template-metafunctions-with-doxygen

**Review by garethellis0 — CHANGES_REQUESTED** [review]
I am _*SUPER*_ happy to see this. Honestly just made my day! :clap: :heart_eyes: Just a few small things, implementation is super clean. We even get more guarantees then with the macro version!

**Review by garethellis0 — APPROVED** [review]
LGTM! :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
Sorry I'm a bit late to the party. This looks great! :+1:

**Pull request merged**